### ISSUE TITLE
Boot (Full) fails if rebooting with a game already running

### ISSUE THREAD

**Issue by deadair3210** [PCSX2/pcsx2#851]
After a reboot, the emulator log will throw a random Syscall Undefined and the virtual PS2 will fail to reboot until reboot is attempted again on which the emulator will either crash or successfully reboot.

Version: PCSX2 1.3.1-20150925184220
CPU: Intel i5-2500k @ 4.0 GHz
GPU: nVidia GeForce GTX 660 Superclocked 3GB
RAM: 11 GB DDR3
Log: https://gist.github.com/deadair3210/a7b941dacc03badcd9e9

**Comment by refractionpcsx2** [comment]
Can you try your game as an ISO and not a CSO, I've had a few games not like being CSO files.

**Comment by avih** [comment]
> Can you try your game as an ISO and not a CSO, I've had a few games not like being CSO files.

This sounds like a bug to me.

Games should not care or be aware that the media is compressed because PCSX2 exposes it the same as uncompressed DVD. Does it also happen with gzip?

This is probably worth filing an issue on it and asking @unknownbrackets to check it out.

**Comment by refractionpcsx2** [comment]
I wish I could remember what games didn't like it, it's going to bug me now :(

**Comment by deadair3210** [comment]
This time it happened with Disgaea Hour of Darkness, but it has happened with other games and without a iso selected before

**Comment by refractionpcsx2** [comment]
Well Disgaea 2 works absolutely fine for me so there's something you're doing that's breaking it or your bios/iso is corrupted.

**Comment by deadair3210** [comment]
Ok, will rerip my BIOS and game and reconvert the game to CSO and see if it fixes it. Will post back here if it continues. Thanks anyway

**Comment by refractionpcsx2** [comment]
Good luck!

**Comment by avih** [comment]
@deadair3210 Not sure why you closed the issue, however, a more detailed and concise description of the issue would help here. It's possible that it works for ref but there's still a bug which you bump into and he doesn't.

Also, I think ref said that CSO compression may _cause_ more bugs, not fix bugs, so if there's indeed a CSO bug then a. it's important that we know about it and b. you should try without CSO compression to see if the bug goes away.

Could you please describe the exact steps you do which end up with the bug?

**Comment by deadair3210** [comment]
1. Select ISO
2. Start Virtual PS2
3. Reboot
4. Random Syscall issue

Going to try to reproduce later, busy for now. Do I need to reopen the issue then?

**Comment by refractionpcsx2** [comment]
What do you mean by reboot? You mean hit "run cdvd" again?

**Comment by deadair3210** [comment]
Yes, sorry. I hit Run CDVD Full so that it goes through the BIOS and then into the game

**Comment by refractionpcsx2** [comment]
Yeh i can't get it to crash doing that :/

**Comment by deadair3210** [comment]
Will try to reproduce it later

**Comment by unknownbrackets** [comment]
@deadair3210 try verifying a crc hash.  If you're using maxcso, it can create a crc hash for you (see the documentation) of either ISO or CSO files.

If the ISO's hash doesn't match a clean hash, then your ISO is dirty.  If the ISO's hash doesn't match the CSO's hash, then maxcso has a bug.  If the hashes are all good, then either pcsx2 has a bug, or the CSO loading code in it does.

Also, what region of game are you trying?  Sometimes, they can behave differently, of course, and have different bugs in emulators.

@refractionpcsx2 can you clarify?  There's been some FUD spread about games "not liking" CSO on PSP due to an implementation bug on physical PSP's custom firmwares.  But obviously, games cannot actually tell.  If there's a bug, let's fix it.

-[Unknown]

**Comment by refractionpcsx2** [comment]
I'll check through my games and see if I can find one and let you know :)

**Comment by deadair3210** [comment]
I'm not sure if this is related, but while testing some other CSO conversions I managed to get the emulator to fail to reset the Virtual PS2. I was testing Dark Cloud, noticed that there were some graphical issues, changed the plugin settings, and hit Run CDVD Full to make sure that the changes would take. After hitting Run, the emulation just stayed on a black screen, and the log was spammed with TLB Miss. Log Attached, sorry the log is so long, had been testing a couple of games before this happened
Log: https://gist.github.com/deadair3210/9e6ca79e8d14819665c7
Sorry if this isnt related enough to be part of this issue, but I wasnt sure if it had anything to do with the Syscall problem considering there are some Unrecoginized op errors around the end of the log as well

**Comment by unknownbrackets** [comment]
Does it happen only when it's a CSO, or does it happen with ISO?

I see right before the corruption there's a log about cheats.  "cheats_ws.zip".  If you disable cheats, does this still happen?

-[Unknown]

**Comment by deadair3210** [comment]
As far as I remember, this only has happened after I converted the game to a CSO, and I'm not sure about the cheats, will try to reproduce with a unconverted copy, though it might take some time as Im not sure what triggered it as it happened after testing about 4-5 other conversions before I started testing Dark Cloud.

**Comment by deadair3210** [comment]
Well, it happened again. This time switching between CSO converted copies of Dynasty Warriors 3 and Dynasty Warriors 4. Dont think the issue is game related, might be format related. Will try reproducing with ISO's. Seems to be triggered by switching between 4-5 games without closing the emulator in between.
![tlb miss](https://cloud.githubusercontent.com/assets/1053504/10125181/a0a61ed2-6530-11e5-9124-6102fcaded8f.PNG)

**Comment by willkuer** [comment]
Full emulog is saved in logs folder in the program directory. But I guess the screenshot is in this case sufficient.

**Comment by deadair3210** [comment]
Just got it to happen with ISO's as well, this time when the PS2 tried to reboot it looped until i shutdown the emulation

**Comment by avih** [comment]
> Select ISO
> Start Virtual PS2
> Reboot
> Random Syscall issue

Just to make sure I understand your procedure:
- Start PCSX2
- Select an ISO from the CDVD menu (CSO?)
- System -> Boot Fast (or full?)
- Play for a while (or try to switch right away?)
- CDVD -> Select a different ISO file (CSO?) (I'm assuming you swap ISO even though your procedure doesn't mention it, is this a correct assumption?)
- PCSX2 prompts to "Swap DISC" or "Reset" -> You choose "Reset".
- Crash ?

**Comment by deadair3210** [comment]
Start PCSX2
Select CSO
System -> Boot Full
Play until the title
CDVD -> Select Different CSO
Swap DISC or Reset -> Reset
Emulated PS2 Crash

**Comment by unknownbrackets** [comment]
But to clarify, it happens even with ISO, and also with multiple games, so presumably the disc or format is irrelevant.

If it's only happening sometimes, perhaps it's a race condition.  Can there be a pending operation that goes uncancelled when you swap discs and reset?

-[Unknown]

**Comment by deadair3210** [comment]
Yes it seems to happen no matter the format or game. It really seems to be either a race condition or a leak somewhere as it only happens after switching games repeatedly

**Comment by avih** [comment]
> Emulated PS2 Crash

You mean PCSX2 crash? or the game doesn't work but PCSX2 doesn't crash?

If PCSX2 doesn't crash, can you still do System -> Boot (fast or full)? Does the game run?

**Comment by deadair3210** [comment]
When i say emulated Ps2 crash i mean that the emulation of the ps2 has ceased functioning correctly, like what happens when a actual ps2 system freezes. the emulator still responds correctly(barring one account) and i can restart the emulation by hitting boot full. I dont use boot fast unless i have to, i like the nostalgic trip of using the ps2 bios

**Comment by avih** [comment]
> the emulator still responds correctly(barring one account) and i can restart the emulation by hitting boot full.

Interesting, so after being prompted to swap disc or reset and choosing reset, it doesn't boot correctly, but then choosing system -> boot full makes it work correctly and the newly selected ISO plays fine?

**Comment by deadair3210** [comment]
yep

**Comment by avih** [comment]
Does it also happen with Boot full (while a game is running) without swapping the CDVD?

**Comment by deadair3210** [comment]
I'm not sure, will try when i get a chance. Computer is busy at the moment

**Comment by avih** [comment]
OK, so I can reproduce. As far as I can tell, it's an issue of boot full while a game is running, either by plainly choosing boot full while a game is running, or by boot full a game, then swap disc, choose "Reset" -> performs boot full (since last boot was full).

I tested with ICO (PAL) and few more. E.g. sequences:
- Start PCSX2
- CDVD -> Choose ICO(PAL)
- Boot fast, wait until the game shows
- Boot fast (OK), wait until the game shows
- Boot fast (OK), wait until the game shows
- Boot full -> doesn't boot correctly

Or
- Start PCSX2, select ICO
- Boot fast (OK), wait till the game shows
- Boot full -> Doesn't boot correctly

Or
- Start PCSX2, select ICO
- Boot full (OK), wait till the game shows
- Boot full -> doesn't boot correctly

**Comment by deadair3210** [comment]
Its good to hear that it wasnt me doing something incorrectly or a problem with my setup.

**Comment by avih** [comment]
Swapping discs after boot fast works OK. After swapping discs it uses the last used moot "mode", so ultimately it's unrelated to swapping discs, it just happens that if the previous boot was full and you swap discs it performs another full boot and the bug shows.

**Comment by avih** [comment]
Anyone wants to find out whether it's a regression, and when it appeared?

**Comment by deadair3210** [comment]
Not sure if this is related, but I noticed that if I switch games the emulator isnt cleaning the buffers for at least Gsdx, as you can see in this picture. I was trying Star Ocean then decided to switch to Tales of Legendia, and noticed you can still see parts of the star ocean menu in the black area for the cutscenes.

Will work on finding when the regression happened, will report back when i do

**Comment by avih** [comment]
> Not sure if this is related

_Probably_ unrelated. Please open a new issue for it. Thanks.

**Comment by deadair3210** [comment]
Cant get the error to happen on PCSX2 1.3.1-20150923092624,think after that is when the regression occured.

Will open a new issue for the Graphical Issue, wasn't sure if it was related to this or not due to the buffers not being cleared somehow interfering with the ps2 rebooting.

**Comment by refractionpcsx2** [comment]
Wild stab in the dark here, but wonder if it was the changes from the coverity-qa extra that modified pcsx2/gui/AppMain.cpp to include the isocdvdsrc (or whatever it was)

**Comment by micove** [comment]
As refraction mentioned https://github.com/PCSX2/pcsx2/commit/4d8a23c1fcedf03773137e5e83ecd6f10ccbbf91 has:

```
 	SysExecEvent_Execute()
 		: m_UseCDVDsrc(false)
 		, m_UseELFOverride(false)
+		, m_cdvdsrc_type(CDVDsrc_Iso)
 	{
 	}
 
 	SysExecEvent_Execute( CDVD_SourceType srctype, const wxString& elf_override )
 		: m_UseCDVDsrc(true)
 		, m_UseELFOverride(true)
 		, m_cdvdsrc_type(srctype)
 		, m_elf_override( elf_override )
 	{
 	}
```

m_UseCDVDsrc is false and m_cdvdsrc_type gets initialized to CDVDsrc_Iso. Can't check this right now.

**Comment by avih** [comment]
> Wild stab in the dark here, but wonder if it was the changes from the coverity-qa extra that modified pcsx2/gui/AppMain.cpp to include the isocdvdsrc (or whatever it was)

I thought so too and then I started testing earlier revisions, going back to January 2015, and all the builds exhibited the same issue.

However, I think my testing procedure conclusions were flawed.

The bug does show consistently for me with ICO and valkyrie profile 2, but doesn't show with any other ISO I currently have on my system. E.g. with Rez or SoTC or GOW2 or MGS3 I can Boot full as many times as I want in a row and it always succeeds.

It appears to me that if either ICO or VP2 are running, then I can't boot full - neither via system -> boot full, nor by swapping the CDVD to any other game and choosing reset (where it tries to perform full boot).

But if another game is running, e.g. SoTC, then I can system -> boot full just fine, and also selecting any other CDVD (even ICO or VP2) and choosing reset -> it succeeds with full boot.

So it looks like some games (my examples are ICO and VP2, and possibly ) corrupt the state of PCSX2 such that it's unable to do a full boot. Fast boot still work though.

To test it from scratch:
- start pcsx2
- choose CDVD GAME_X
- boot fast (or full - doesn't matter)
- wait till you see ingame stuff.
- boot full <-- if this doesn't work, then GAME_X is one of those problematic games.

**Comment by avih** [comment]
It's possible that other games are affected too, but it depends on whether or not they've reached the code/state where PCSX2 gets corrupted.

With VP2 and ICO - it's enough to see the first ingame thingy* such that boot full no longer works.

* In VP2 it's the screen setup, in ICO it's the language selection (assuming no memory card).

**Comment by deadair3210** [comment]
Found a game that does this every time i have tried.
Load Kingdom Hearts 2 Final Mix (NTSC-J), Boot(Full), Wait until the PS2 logo is about to pop up and hit Boot(Full) again, Syscall Undefined
Log - https://gist.github.com/deadair3210/63485a69ed5c5ae70e2e

**Comment by deadair3210** [comment]
Same thing happens with Kingdom Hearts Final Mix (NTSC-J), same procedure same results

**Comment by deadair3210** [comment]
Went from Soulcaliber 3's main menu, switched discs to Sphinx and the Cursed Mummy, Syscall

**Comment by gregory38** [comment]
Is it a regession or is it a new bug? Status is quite confusing

**Comment by jcdenton2k** [comment]
@deadair3210 I've taken the time to look through your log and noticed a few odd things.

1 - For whatever reason you still insist on using the CSO format despite the fact that this increases loading times in a performance-heavy emulator like PCSX2.  It takes more time to read the data (and some of it has to be decompressed on the fly) and as a result it can cause performance issues.  I don't know why even in 2015 or now in 2016 you insist on trading hard drive space (which is dirt-cheap) in favor of a performance compromise.

2 - This part:

> Loading patch 'F266B00B.pnach' from archive 'X:\Emulation\Emulators\Pcsx2\pcsx2-v1.3.1-1198-gc700035-windows-x86\pcsx2-v1.3.1-1198-gc700035-windows-x86\cheats_ws.zip'
comment: Kingdom Hearts 2 Final Mix + (NTSC-J)
(Wide Screen Cheats DB) Patches Loaded: 17

Under the 'System' menu of PCSX2, make sure to NOT have a checkmark under 'Enable Cheats', 'Enable Widescreen Patches', and 'Enable Host Filesystem'.  You can use the 'Automatic Gamefixes' if you like but you can try to disable that and see if the issue still happens.

**Comment by unknownbrackets** [comment]
Well, unless you don't have enough RAM for your OS to fully cache the entire ISO in memory + the emulator and its memory usage, but you do have the RAM to cache the entire CSO.  There can be lots of trade offs.

-[Unknown]

**Comment by jcdenton2k** [comment]
@deadair3210 @gregory38 
I've found a solution to this issue so that it can be closed.

Using PCSX2 x32 build v1.5.0-dev-1010-gd58c07d compiled July-22-2016 by Orphis Buildbot from the official PCSX2 download links.

Using Win7 x64 OS and Windows build of PCSX2.

Initially I was able to reproduce the above issue with the TLB-miss and errors.  Then I did a bit of configuration changing to see what was up.

I narrowed down the cause of the issue to enabling 'Widescreen Patches'.  If you keep them disabled then the game will load fine regardless of how many times you 'Reload Full' from the BIOS.

**Comment by deadair3210** [comment]
I use CSO because I have 11 GB of RAM so the image is cached in its entirety, and would rather be able to have more room on my drive for other games than have to worry about the (negligible) perf loss of having to decompress on the fly, which doesnt take that much on my i5-2500k. I have a 350 GB drive dedicated to all my emulation and code projects

**Comment by jcdenton2k** [comment]
I opened up the file and here were the ones present in my version:

> // new 16:9 hack by - ElHecht
patch=1,EE,00106e10,word,3c013f4c // c480004c
patch=1,EE,00106e28,word,3421cccc // 4600a7c6
patch=1,EE,00106e2c,word,4481f800 // 00000000
patch=1,EE,00106e30,word,461fa503 // 4600a503
patch=1,EE,00106e54,word,3c1b3f40 // 00000000 hor fov
patch=1,EE,00106e58,word,449bf000 // 00000000
patch=1,EE,00106e70,word,461effc2 // 00000000
patch=1,EE,00106e74,word,e61f004c // 00000000

// optional zoom for cutscenes
patch=1,EE,2036a0bc,word,43f90000 // 43d00000

//font fix
patch=1,EE,2036CE94,word,3F400000 // 3F800000
patch=1,EE,2036CE98,word,3F400000 // 3F800000
patch=1,EE,2036CE9C,word,3F400000 // 3F800000

//black border fix
patch=1,EE,0014AD80,word,24050000
patch=1,EE,0014ADA8,word,24050000
patch=1,EE,0014ADD0,word,24050000
patch=1,EE,0014AE00,word,24050000

//lower subtitles
patch=1,EE,001ac8d8,word,240a018a

So I'm not sure which of those is the cause of this issue.  Starting PCSX2 fresh (or just Fast Boot) is the easiest solution/workaround.

**Comment by deadair3210** [comment]
I don't have access to my computer currently, or I would test this and attempt to help diagnose more.

**Comment by jcdenton2k** [comment]
Since this issue is a widescreen patch issue, your best bet is to try poking the person responsible for those and passing along a link to this thread.  Let them handle it while you work with them to narrow down which code(s) are doing the borking.

As I've said earlier, it may just be easier to stick a comment in the patch noting this issue and let that be that.

**Comment by deadair3210** [comment]
Unfortunatly, I am unaware of who is in charge of the widescreen patches, other than it at one point was handled by the forums and then updated at certain points. I swear at some point I tried it without the patches enabled, and still had the issues. The issue originally arised while testing and running an English translation of Kingdom Hearts Final Mix, which makes me wonder if its not a bad interaction with the patches, and the NTSC-J video mode that Final Mix uses

**Comment by jcdenton2k** [comment]
NTSC = 60 Hz = same video mode as other NTSC games.  NTSC is the standard for most games.  Doesn't matter if it says NTSC-USA or NTSC-Japan.

PAL = 50 Hz = the competing standard primarily in European markets.  I feel bad for them being stuck at 50 FPS instead of 60.

**Comment by deadair3210** [comment]
I was unaware if there was a difference between J and U/C. Just shooting theories as I think about them. What exactly is a TLB Miss, knowing that might help

-----Original Message-----

**Comment by jcdenton2k** [comment]
Maybe one of the developers can explain 'TLB Miss' better than I can as I'm not a coder.
It typically happens with a bad ISO rip and it is suggested to use ImgBurn when ripping ISOs to make sure they work properly.

In the case of this issue, the TLB Miss happens because of some conflict between a restart of the emulator and one or more of the widescreen patches and/or cheats for specific games.

**Comment by Sarania** [comment]
There is a difference between TLB miss [load] and TLB miss [store]. [load] is generally caused by bad ISO whereas [store] is often caused by bad codes/patches. Basically TLB is how the PS2(and other machines) deal with virtual memory. When the game asks for data at a location in virtual memory space, the TLB translates that back to physical address space. However, if the address is invalid it can "miss" i.e. point to an invalid space. [load] and [store] just signify whether it was a read or a write.

https://en.wikipedia.org/wiki/Translation_lookaside_buffer

**Comment by ramapcsx2** [comment]
Even if it's caused by patches, this is still a valid problem. "Rebooting" in PCSX2 is supposed to clean the slate and forget everything about the emulation state. Maybe a bug in the patching code deinit?

**Comment by deadair3210** [comment]
From what I remember, and what I have gathered, the issue happens after rebooting from any game that has patches. That would explain why the issue was different depending on wether I was running the English translation of Final Mix, as the CRC changed so the patches no longer applied

**Comment by avih** [comment]
> I've found a solution to this issue so that it can be closed.

No, it can't be closed.

The fact that you found a "solution" means there's a problem, hence the bug should be kept open either until the problem is fixed or until the PCSX2 team decided not to fix it. Thus far none of those has happened.

**Comment by avih** [comment]
Also, I think I can confirm that at least with Ico, the issue is indeed the Widescreen patches enabled (Ico does have widescreen patches). I.e. If I have Ico (PAL) running and the ws patches are enabled (regardless if the automatic game fixes are enabled or not - and there are 2 automatic configuration fixes for Ico), and then reboot-full then it fails.

But If Ico is running and I don't have the widescreen patches enabled (again, regardless if the automatic fixes are enabled or not), then I can reboot-full and it works.

So I tried to disable the Widescreen patches from the menu, and instead put the widescreen patch (`5C991F4E.pnach` from `cheats_ws.zip`) at the `cheats` folder - and had the same issue egain - if cheats are enabled and the patch is loaded, then boot-full fails.

So it's _not_ related specifically to the widescreen patches "system" in pcsx2, but rather to whether or not (some kinds of?) patches are loaded - if they're loaded, e.g. when `Enable Cheats` is checked at the menu and the patch exists at the `cheats` folder - then reboot-full fails.

**Comment by avih** [comment]
We can probably disable all patches when the user requests a full boot and restore them after the boot starts, but that would be a workaround.

I'm guessing the real issue is incorrect or insufficient shutdown of the VM when patches are loaded, and the real fix would therefore probably be at the shutdown procedure.

**Comment by Aced14** [comment]
This issue and #1310 are likely caused by the same bug. Ironically, it benefits #627 when repeatedly fast-booting.

This isn't a proper solution, but pnaches can be edited to work around the bug. If an [E-type (multi-line skip) code](https://web.archive.org/web/20140304105930/http://www.codemasters-project.net/guides/showentry.php?e=847&epage=16) is used, hacks can be activated only when they're supposed to. For example, an E-type code can be used to detect whether a specific memory address contains an expected value to conditionally enable all of a pnach's hacks. That way, if a user attempts to full boot while a game is already running, the E-type code will prevent the previously-running game's pnach hacks from "polluting" BIOS memory.

Many of the pnaches I've contributed to the widescreen archive are already using E-type codes to work around this bug (and another unrelated full boot pnach bug).

**Comment by avih** [comment]
> This isn't a proper solution, but pnaches can be edited to work around the bug ...

Right, it's not a good solution.

I tracked this a bit, and it seems the issue is that we apply the patches/cheats for the previous CRC when rebooting - to the bios on full boot, and not sure whether or not to the new elf on fast boot. And this can break the bios.

This suggests that either we don't reset the CRC on [re]boot soon enough, or we do but in the grand scheme of things it doesn't propagate fast enough to where the patches are applied (seemingly at-least/also at `SysCoreThread::VsyncInThread()`).

But the CRC (`elfCRC`) does reset on time correctly, and `SysCoreThread::VsyncInThread()` doesn't care at all about the current CRC - it just applies whatever patches we already have initialized/loaded.

So now the issue seems that we don't unload/invalidate the current patches when rebooting.

I experimented with two approaches, and both worked for the specific issue of full boot:
1. Either: at `SysCoreThread::VsyncInThread()` add `if (!ElfCRC) return;` - so this just doesn't apply the patches while we're at the bios or before the crc is known. It works, but IMO not great.
2. Or: at `SysCoreThread::_reset_stuff_as_needed()` - if `m_resetVirtualMachine` then also `ResetCheatsCount();`.

2 seems better and does fix the issue, but actually it's only partial, since it resets the cheats (and widescreen hacks which internally are cheats too), but doesn't reset the patches count (the ones applied from the games DB). So we should probably reset the patches too.

While at it, I noticed that some places do:
```
if (EmuConfig.EnablePatches) ApplyPatch();
if (EmuConfig.EnableWideScreenPatches)  ApplyCheat();
```
But ignore `EmuConfig.EnableWideScreenPatches`.

This happens at `SysCoreThread::GameStartingInThread()` and at `iR5900-32.cpp::recRecompile()` (but at `SysCoreThread::VsyncInThread()` it also checks `EmuConfig.EnableWideScreenPatches`).

The tests are actually almost* moot, since on every config change (or boot) the patches and cheats are reset and loaded from scratch, so there's no need to check whether the config enables them or not, as if it doesn't then they won't be loaded to begin with (* almost - since it incorrectly prevents loading the widescreen hacks unless cheats are enabled).

Overal, IMO we should do the following:

1. Most importantly - invalidate the currently loaded patches and cheats when rebooting. This should fix this bug, and should be relatively easy.

2. Unify patches/[widescreen]cheats handling, where the only difference would be whether or not we load them - according to the config (game db patches, "cheats" AKA user patches, and widescreen hacks patches) - the load part already works at the current code. Other than loading them on boot/config-changes, the rest of pcsx2 will have only two visible functions which handles all of them: `invalidate` (used for reboot and config changes before [re]-loading them) and `apply` (on vsync, recompile, etc, whenever they need to be applied).

**Comment by avih** [comment]
@refractionpcsx2 @ramapcsx2  @Sarania @prafullpcsx2 @Aced14 (or anyone else who has a good answer)

We have patch lines which start with `0,...` e.g.:
`patch=0,EE,0010BC88,word,48468800`
and also lines which start with `1,...` e.g.:
`patch=1,EE,001110e0,word,00000000`

E.g. at `GameIndex.dbf` we have 240 patch lines which start with 0, and 45 which start with 1 (didn't count how many of them are commented out, but we do have both 0 and 1 uncommented). At `cheats_ws.zip` I checked few random files, and all of them only had lines starting with 1.

I can see what difference it makes inside PCSX2, but I'm not seeing anything which documents how it _should_ work for 0 and 1.

So my questions to you are:
- Do you know what it _should_ do for 0 and 1?
- How do you know? (is it part of some standard? is it PCSX2 specific? is it "common knowledge"?)
- Some links to where it's documented or mentioned? (related to PCSX2 or elsewhere)?

Thanks.

**Comment by ramapcsx2** [comment]
Oh, this is really old stuff. If I remember correctly, one was for "overwrite once", the other for "overwrite always". I could be wrong and I can't even tell which was which :p

**Comment by avih** [comment]
@deadair3210 @Aced14 (or others), could you please confirm (or deny) that this is now fixed with the recent commits?

**Comment by deadair3210** [comment]
Unfortunatly, I am unable to test this at any close time as my computer is not where I am

**Comment by Sarania** [comment]
Just FYI (since you did ping me) I will confirm that what @ramapcsx2 said is also the explanation I've heard between patch=0 and patch=1. I think it's almost become forgotten knowledge nowadays as I haven't seen it brought up in a while! I _think_ it was 0 was overwrite once and 1 was overwrite always. I could be wrong though. I think nowadays people are just taught to use =1 since in all but a few specific cases it should be the "safe" way(if I'm write about it being "overwrite always").

**Comment by avih** [comment]
Right, that's what the code does - applies patches with "0" once on startup, and patches with "1" on every vsync. I already added docs for this at `Patch.h` and elsewhere, and also gave this type a name `patch_place_type` with two values `PTT_ONCE_ON_LOAD` and `PPT_CONTINUOUSLY`.

Just for the record, @gigaherz and @ramapcsx2 agreed that the pnach format was invented by nachbrenner (a past PCSX2 developer), and hence the format name too ;)

Apparently it was common knowledge back then what do the 0/1 do, however we couldn't find even a single comment at the code or elsewhere which describes what it _should_ do, which leaves us with a definition which follows the implementation.

As mentioned earlier, ~90% of the patches at `GameIndex.dbf` use 0 - to apply the patch once on startup. Apparently typically that's enough.

Off topic, well.. not really.. if there are still issues with fast or full reboot where patches are applied correctly or incorrectly, let me know. I played with it quite a bit with many reboots, enabling/disabling various types of patches/cheats/etc, changing games, rebooting full or fast, and as far as I can tell it now always works correctly.

**Comment by Aced14** [comment]
@avih Thanks :)!

I came up with some fast boot to full boot tests using widescreen patches that caused BIOS crashes/log errors in an older recent build (56adb85a87c763f078b57f395cdf8fc60917ca32). I then built 56adb85a87c763f078b57f395cdf8fc60917ca32 in VS 2013 and retried the same tests... worked great :). cff8cb137cc3f1f73096d54cbecc6d79cd0781ff didn't have any problems either.

My tests did crash at a later point in time, but it's unrelated to this issue. I'll post a separate issue about it later today or tomorrow.
